src/plugins/media/media.js

```diff
@@ -0,0 +1,172 @@
+  };
+  
+  onPause = function (event) {
+    var type = event.target.nodeName.toLowerCase();
+    document.body.classList.add(type + "paused");
+    document.body.classList.remove(type + "playing");
+  };
+  
+  onEnded = function (event) {
+    var type = event.target.nodeName.toLowerCase();
+    document.body.classList.remove(type + "playing");
+    document.body.classList.remove(type + "paused");
+  };
+  
+  enhanceMediaNode = function (type) {
+    var i, id, media = document.getElementsByTagName(type);
+    for (i = 0; i < media.length; i++) {
+      // Set an id to identify each media node - used e.g. by the consoleMedia plugin
+      id = media[i].getAttribute('id');
+      if (id === undefined || id === null) {
+        media[i].setAttribute('id', 'media-' + type + '-' + i);
+      }
+      media[i].addEventListener("play", onPlay, false);
+      media[i].addEventListener("playing", onPlay, false);
+      media[i].addEventListener("pause", onPause, false);
+      media[i].addEventListener("ended", onEnded, false);
+    }
+  };
+  
+  enhanceMedia = function () {
+    var i, steps, media = document.getElementsByClassName('media');
+    for (i = 0; i < media.length; i++) {
+      parseMediaShortcut(media[i]);
+    }
+    enhanceMediaNode('audio');
+    enhanceMediaNode('video');
+    steps = document.getElementsByClassName('step');
+    for (i = 0; i < steps.length; i++) {
+      console.log("Add event listener.");
+      steps[i].addEventListener('impress:stepenter', playOnStepenter, false);
+    }
+  };
+  
+  playOnStepenter = function (event) {
+    console.log("Stepenter triggered.");
+    var videos, audios, media, play, type, i;
+    if ((!event) || (!event.target)) {
+      return;
+    }
+    
+    audios = event.target.getElementsByTagName('audio');
+    videos = event.target.getElementsByTagName('video');
+    media = {
+      video: videos,
+      audio: audios
+    };
+    for (type in media) {
+      for (i = 0; i < media[type].length; i++) {
+        play = media[type][i].dataset.mediaPlayonenter;
+        console.log(play);
+        if (play !== undefined) {
+          media[type][i].play();
+        }
+      }
+    }
+  }
+
+  pauseOnStepleave = function (event) {
+    var videos, audios, media, type, i;
+    if ((!event) || (!event.target)) {
+      return;
+    }
+
+    videos = event.target.getElementsByTagName('video');
+    audios = event.target.getElementsByTagName('audio');
+    media = {
+      video: videos,
+      audio: audios
+    };
+    for (type in media) {
+      for (i = 0; i < media[type].length; i++) {
+        if (media[type][i].dataset.mediaPauseonleave !== "false") {
+          media[type][i].pause();
+        }
+      }
+    }
+  };
+
+  // Register the plugin to be called in the pre-init phase
+  window.impress.addPreInitPlugin(enhanceMedia);
+
+  // Register the plugin to be called in pre-stepleave phase
+  impress.addPreStepLeavePlugin(pauseOnStepleave);
```

Comment:
The `pauseOnStepleave` (or equivalent, whatever we end up calling it) should be called on `impress:stepleave`. Just like `playOnStepenter` is triggered on `impress:stepenter`. It has no need to be a pre-stepleave plugin.

Reply:
What exactly is the difference between a pre-stepleave plugin and binding it to `impress:stepleave`? I was already wondering, why there is no pre-stepenter plugin.

Reply:
The original idea for impress.js events is based on using javascript events. This is the preferred mechanism whenever suitable.

The pre- style events were added later to facilitate two use cases:
* Cases where the plugin really needs to run before the event it is tied to. For example, the rel plugin clearly must run before init itself.
* Cases where you may want to prevent the event from happening at all. For example, the stop plugin returns false, which stops the stepleave activity completely. (You will stay on the current step and not move to next.)

Reply:
Understood :-)